File: supplementary-materials/langchain-vs-springai/backend/python/CsvIeltsTask2Loader.py
import csv
import os
import time
from pathlib import Path
from typing import Any
from langchain_community.vectorstores import FAISS
from langchain_core.documents import Document
from langchain_openai import OpenAIEmbeddings

class CsvIeltsTask2Loader:
    def __init__(self, csv_path: str, vector_store_path: str, preprocessor: Any):
        self.csv_path = csv_path
        self.vector_store_path = Path(vector_store_path)
        self.preprocessor = preprocessor
        self.batch_size = 100
        self.delay_seconds = 5
        self.embeddings = OpenAIEmbeddings()

    def load_csv_essays(self):
        '''Loads essays from a CSV file, processes them, and stores them in a vector store.
        Raises:
            FileNotFoundError: If the CSV file does not exist.
            ValueError: If no valid documents are found in the CSV.
        '''

        # Check if the CSV file exists
        if not os.path.exists(self.csv_path):
            raise FileNotFoundError(f"CSV file not found at: {self.csv_path}")
        
        # Check if the vector store is already loaded
        if self._is_vector_store_loaded():
            print("Vector store already loaded, skipping CSV processing.")
            return

        documents = []

        with open(self.csv_path, newline='', encoding='utf-8') as csvfile:
            reader = csv.reader(csvfile)
            header = next(reader)
            print(f"CSV Header: {header}")

            processed_count = 0
            skipped_count = 0

            for line_num, row in enumerate(reader, 2):
                try:
                    if len(row) < 9:
                        print(f"Line {line_num}: Skipped - Only {len(row)} columns found")
                        skipped_count += 1
                        continue

                    # Check if the task type is 2 (filter the dataset)
                    if row[0].strip().lower() != "2":
                        skipped_count += 1
                        continue

                    question = row[1].strip()
                    raw_essay = row[2].strip()
                    overall_score = row[8].strip()

                    if not all([question, raw_essay, overall_score]):
                        print(f"Line {line_num}: Skipped - Missing required fields")
                        skipped_count += 1
                        continue

                    clean_essay = self.preprocessor.clean_essay(raw_essay)
                    topic = self.preprocessor.extract_main_topic(question)
                    word_count = self.preprocessor.count_words(clean_essay)

                    content = self._build_document_content(
                        question, clean_essay, row[3], overall_score
                    )

                    metadata = {
                        "type": "task2_essay",
                        "band": overall_score,
                        "question": question,
                        "topic": topic,
                        "word_count": word_count,
                        "source_line": line_num
                    }

                    documents.append(Document(page_content=content, metadata=metadata))
                    processed_count += 1

                except Exception as e:
                    print(f"Error processing line {line_num}: {str(e)}")
                    skipped_count += 1

        print(f"Processing Summary:\n - Processed: {processed_count}\n - Skipped: {skipped_count}")

        if not documents:
            raise ValueError("No valid documents found in CSV")

        # Init FAISS
        vector_store = FAISS.from_documents(documents[:self.batch_size], self.embeddings)
        remaining = documents[self.batch_size:]

        # Process in batches
        for i in range(0, len(remaining), self.batch_size):
            batch = remaining[i:i+self.batch_size]
            vector_store.add_documents(batch)
            vector_store.save_local(self.vector_store_path)

            print(f"Processed {i + self.batch_size + len(batch)}/{len(documents)} documents")

            time.sleep(self.delay_seconds)


        # Save the vector store
        vector_store.save_local(self.vector_store_path)



    def _is_vector_store_loaded(self) -> bool:
        '''Checks if the vector store is already loaded by looking for the index file.
        Returns:
            bool: True if the vector store is loaded, False otherwise.
        '''
        return (self.vector_store_path / "index.faiss").exists()
    


    def _build_document_content(self, question: str, essay: str, examiner_comment: str, score: str) -> str:
        '''Builds the content for the document to be stored in the vector store.
        Args:
            question (str): The essay question.
            essay (str): The essay text.
            examiner_comment (str): Comments from the examiner.
            score (str): The band score.
        Returns:
            str: The formatted content string.
        '''

        content = f"IELTS Writing Task 2 Essay (Band {score})\n\nQuestion:\n{question}\n\nEssay:\n{essay}\n\n"
        if examiner_comment.strip():
            content += f"Examiner Comments:\n{examiner_comment}\n\n"
        content += f"Scores:\n- Overall: {score}"
        return content

File: supplementary-materials/langchain-vs-springai/backend/java/src/main/java/org/nc/IELTSChecker/services/CsvIeltsTask2Loader.java
package org.nc.IELTSChecker.services;

import com.opencsv.CSVReader;
import org.nc.IELTSChecker.model.EssayDocument;
import org.springframework.ai.document.Document;
import org.springframework.ai.vectorstore.SimpleVectorStore;
import org.springframework.ai.vectorstore.VectorStore;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.io.Resource;
import org.springframework.stereotype.Service;

import java.io.File;
import java.io.InputStreamReader;
import java.util.*;

@Service
public class CsvIeltsTask2Loader {

    @Value("classpath:/data/ielts_writing_dataset.csv")
    private Resource csvFile;

    private static final int BATCH_SIZE = 100;
    private static final int DELAY_MS = 5000;

    @Autowired
    private File vectorStoreFile;

    @Autowired
    private VectorStore vectorStore;

    @Autowired
    private EssayPreprocessor preprocessor;


    /**
     * Loads essays from a CSV file and processes them into a vector store.
     * @throws Exception If an error occurs during file reading or processing.
     */
    public void loadCsvEssays() throws Exception {
        List<Document> documents = new ArrayList<>();
        System.out.println("Starting IELTS Dataset loading process...");

        if (!csvFile.exists()) {
            throw new IllegalStateException("CSV file not found at: " + csvFile.getURI());
        }
        if(isVectorStoreAlreadyLoaded()) {
            System.out.println("Vector store already loaded, skipping CSV processing.");
            return;
        }

        try (CSVReader reader = new CSVReader(new InputStreamReader(csvFile.getInputStream()))) {
            String[] header = reader.readNext();
            System.out.println("CSV Header: " + Arrays.toString(header));

            int lineNumber = 1;
            int processedCount = 0;
            int skippedCount = 0;

            String[] line;
            while ((line = reader.readNext()) != null) {
                lineNumber++;
                try {
                    if (line.length < 9) {
                        System.out.printf("Line %d: Skipped - Only %d columns found%n", lineNumber, line.length);
                        skippedCount++;
                        continue;
                    }

                    // Only process Task 2 essays
                    String taskType = line[0].trim();
                    if (!taskType.equalsIgnoreCase("2")) {
                        skippedCount++;
                        continue;
                    }

                    // Validate fields
                    String question = line[1].trim();
                    String rawEssay = line[2].trim();
                    String overallScore = line[8].trim();

                    if (question.isEmpty() || rawEssay.isEmpty() || overallScore.isEmpty()) {
                        System.out.printf("Line %d: Skipped - Missing required fields%n", lineNumber);
                        skippedCount++;
                        continue;
                    }

                    String cleanEssay = preprocessor.cleanEssay(rawEssay);
                    String topic = preprocessor.extractMainTopic(question);
                    int wordCount = preprocessor.countWords(cleanEssay);

                    String content = buildDocumentContent(
                            question,
                            cleanEssay,
                            line[3],    // Examiner comment
                            overallScore
                    );

                    // Create metadata
                    Map<String, Object> metadata = new HashMap<>();
                    metadata.put("type", "task2_essay");
                    metadata.put("band", overallScore);
                    metadata.put("question", question);
                    metadata.put("topic", topic);
                    metadata.put("word_count", wordCount);
                    metadata.put("source_line", lineNumber);

                    // Create and store document
                    documents.add(new EssayDocument(content, metadata, overallScore, question, topic));
                    processedCount++;

                } catch (Exception e) {
                    System.err.printf("Error processing line %d: %s%n", lineNumber, e.getMessage());
                    System.err.println("Problematic line: " + Arrays.toString(line));
                    skippedCount++;
                }
            }

            System.out.printf("Processing Summary:%n");
            System.out.printf(" - Total lines processed: %d%n", lineNumber - 1);
            System.out.printf(" - Successfully processed: %d%n", processedCount);
            System.out.printf(" - Skipped lines: %d%n", skippedCount);


            if (documents.isEmpty()) throw new IllegalStateException("Error: No valid documents found in the CSV file.");

            // Process documents in batches
            System.out.println("Processing documents in batches...");
            int total = documents.size();
            for (int i = 0; i < total; i += BATCH_SIZE) {
                int end = Math.min(i + BATCH_SIZE, total);
                List<Document> batch = documents.subList(i, end);

                vectorStore.accept(batch);
                saveVectorStore();

                System.out.printf("Processed %d/%d documents (%.1f%%)%n",
                        end, total, (end * 100.0) / total);

                Thread.sleep(DELAY_MS);
            }
        } catch (Exception e) {
            System.err.println("Critical error during CSV processing: " + e.getMessage());
            throw e;
        }
    }

    /**
     * Builds the content string for the essay document.
     * @param question
     * @param essay
     * @param examinerComment
     * @param overallScore
     * @return Returns the formatted content string.
     */
    private String buildDocumentContent(String question, String essay, String examinerComment, String overallScore) {

            StringBuilder contentBuilder = new StringBuilder();
            contentBuilder.append(String.format(
                    "IELTS Writing Task 2 Essay (Band %s)%n%nQuestion:%n%s%n%nEssay:%n%s%n%n",
                    overallScore, question, essay
            ));

            if (!examinerComment.isEmpty())
                contentBuilder.append(String.format("Examiner Comments:%n%s%n%n", examinerComment));

            contentBuilder.append(String.format("Scores:%n- Overall: %s", overallScore));

            return contentBuilder.toString();
        }


    /**
     * Saves the vector store to a file.
     */
    private void saveVectorStore() {
        if (!(vectorStore instanceof SimpleVectorStore))
            throw new IllegalStateException("Vector store is misconfigured. Expected SimpleVectorStore.");


        try {
            ((SimpleVectorStore) vectorStore).save(vectorStoreFile);
            System.out.println("Saved vector store to: " + vectorStoreFile.getAbsolutePath());
        } catch (Exception e) {
            System.err.println("Error saving vector store: " + e.getMessage());
            throw new RuntimeException("Vector store persistence failed", e);
        }
    }

    /**
     * Checks if the vector store is already loaded.
     * @return Returns true if the vector store file exists and is not empty.
     */
    public boolean isVectorStoreAlreadyLoaded() {
        return vectorStoreFile.exists() && vectorStoreFile.length() > 0;
    }
}